### PR TITLE
Fault Tolerant Manual: Add stateful dataloader iter

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -441,7 +449,9 @@ def _find_current_worker(iterator: Iterator) -> Dict[str, Optional[int]]:
     return {"num_workers": num_workers, "previous_worker": previous_worker}
 
 
-def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate: Callable) -> Dict:
+def _capture_metadata_collate(
+    samples: List, dataset: Dataset, default_collate: Callable, fault_tolerant_mode: _FaultTolerantMode
+) -> Dict:
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:454]: the return type is not accurate due to the early return path

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -441,7 +449,9 @@ def _find_current_worker(iterator: Iterator) -> Dict[str, Optional[int]]:
     return {"num_workers": num_workers, "previous_worker": previous_worker}
 
 
-def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate: Callable) -> Dict:
+def _capture_metadata_collate(
+    samples: List, dataset: Dataset, default_collate: Callable, fault_tolerant_mode: _FaultTolerantMode
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:453]:
```suggestion
    samples: List, dataset: Dataset, collate: Callable, fault_tolerant_mode: _FaultTolerantMode
```

this is not necessarily the default collate

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -570,3 +602,126 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _is_obj_stateful(obj: Any) -> bool:
+    """In order to be stateful, an object should implement a ``state_dict`` and ``load_state_dict`` method."""
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:608]: If you say it should implement both, we also should check for both methods

File: pytorch_lightning/utilities/auto_restart.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:453]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -454,10 +461,26 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
             "__pl_restart_meta": {"sampler_name0": state_dict0, "sampler_name1": state_dict1},
         }
     """
-    data = default_collate(samples)
-    if not isinstance(dataset, (CaptureIterableDataset, CaptureMapDataset)):
+    data = collate(samples)
+    fault_tolerant_mode
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:465]: @tchaton accidentally placed?

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -580,3 +609,116 @@ def state_dict(self) -> Dict[str, Any]:
...
+                num_batches_fetched=self.num_batches_fetched,
+            )
+        ]
+        # ensures there is an alignement between the sampler state and currently fetched batch
+        assert sampler_state_idx == self.num_batches_fetched
+        self._data_fetcher._store_dataloader_iter_state(self, state)
+        return batch
+
+
+class _SingleProcessDataLoaderIterStateful(_StatefulMixin, _SingleProcessDataLoaderIter):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:686]: "StatefulMixin" is not a mixin if the implementation of this class relies on it. Currently, what I recognize here is multiple inheritance for the purpose of code reuse. So let's not call it a mixin?

File: tests/utilities/test_auto_restart.py [hunk 6]
```diff
@@ -1237,3 +1241,109 @@ def test_fault_tolerant_mode_enum():
...
+
+    def state_dict(self):
+        self.counter += 1
+        return {"counter": self.counter}
+
+    def load_state_dict(self, state_dict):
+        self.counter = state_dict["counter"]
+
+
+class FailingStatefulRandomDataset(RandomDataset):
```
Comment (awaelchli) [inline tests/utilities/test_auto_restart.py:1258]: what makes this failing?

File: tests/utilities/test_auto_restart.py [hunk 7]
```diff
@@ -1237,3 +1241,109 @@ def test_fault_tolerant_mode_enum():
...
+
+
+class StatefulRandomDataset(FailingStatefulRandomDataset):
+    def state_dict(self):
+        info = get_worker_info()
+        worker_id = info.id if info else 0
+        return {worker_id: {"counter": self.counter}}
+
+
+@pytest.mark.parametrize("num_workers", [0, 2])
```
Comment (awaelchli) [inline tests/utilities/test_auto_restart.py:1281]: will you exclude num_workers=2 before merging?

File: pytorch_lightning/utilities/auto_restart.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py]: Done !

File: tests/utilities/test_auto_restart.py (hunk 6 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py]: Cleaned up !

File: tests/utilities/test_auto_restart.py (hunk 7 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:1277]: It takes 3 sec in total. I would prefer to include it as it covers the `_MultiProcessingDataLoaderIterStateful` functionalities.

Comment (awaelchli) [reply tests/utilities/test_auto_restart.py:1277]: I know, but unfortunately 3 secs are an eternity considering we have thousands of tests. Recently we have hit our CI timeout of 40 minutes! cc @carmocca

Comment (tchaton) [reply tests/utilities/test_auto_restart.py:1277]: Fine for now, but let's re-explore the others auto restart tests. Some are taking 20 secs.

Comment (tchaton) [reply tests/utilities/test_auto_restart.py]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 8]
```diff
@@ -447,10 +455,25 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
-    metadata = dataset.state_dict()
+    data = collate_fn(samples)
+    metadata = None
+    if fault_tolerant_mode.is_automatic:
+        metadata = dataset.state_dict()
+    else:
+        state_dict_fn = getattr(dataset, "state_dict", None)
+        info = get_worker_info()
+        worker_id = info.id if info else 0
+        if state_dict_fn:
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/auto_restart.py:466]: Better explicit here:
```suggestion
        if state_dict_fn is not None:
```

File: pytorch_lightning/utilities/auto_restart.py [hunk 9]
```diff
@@ -589,3 +621,116 @@ def state_dict(self) -> Dict[str, Any]:
...
+    def _reset(self, loader: DataLoader, first_iter: bool = False):
+        super()._reset(loader, first_iter=first_iter)
+        self._loader = loader
+        self.num_batches_fetched = 0
+
+    def __accumulate_state(self, sampler_state: Dict[str, Any]) -> None:
+        # initialize the queue if it doesn't exist.
+        if not hasattr(self, "_sampler_state"):
+            self._sampler_state = []
+            self._sampler_state_idx = 0
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/auto_restart.py:638]: Can be removed as this will be handled by the next line:
```suggestion
```

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/trainer/data_loading.py [hunk 10]
```diff
@@ -441,9 +440,7 @@ def request_dataloader(
     def _add_sampler_metadata_collate(dataloader: DataLoader) -> None:
         """Wrap default collate function to retrive ``FastForwardSampler`` state dict when fault tolerant is
         enabled."""
-        dataloader.collate_fn = partial(
-            _capture_metadata_collate, dataset=dataloader.dataset, default_collate=dataloader.collate_fn
-        )
+        _add_capture_metadata_collate(dataloader)
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:443]: If `_add_sampler_metadata_collate` only calls `_add_capture_metadata_collate`, shouldn't all occurrences of the former be replaced with the latter?

File: pytorch_lightning/utilities/auto_restart.py [hunk 11]
```diff
@@ -480,6 +503,9 @@ def patch_dataloader_iterator(
     will extract the current iteration as part of the metadata returned by a custom batch.
     """
 
+    if not _FaultTolerantMode.detect_current_mode().is_automatic:
+        return
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:507]: Shouldn't this be checked wherever this gets called?

File: pytorch_lightning/utilities/auto_restart.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:466]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:638]: Done !

File: pytorch_lightning/trainer/data_loading.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:443]: Yes, entirely right !

File: pytorch_lightning/utilities/auto_restart.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:507]: Not sure to fully understand your question. Yes, it needs to be checked every time it gets called. But this code will be refactored soon.

File: pytorch_lightning/utilities/auto_restart.py [hunk 12]
```diff
@@ -447,10 +455,25 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
+            if worker_id not in metadata:
+                if info and info.num_workers > 1:
+                    raise MisconfigurationException(
+                        f"The state_dict returned by {dataset} needs to be indexed by `worker_id` integer keys."
+                    )
+                metadata = {0: metadata}
+        if metadata is None:
+            metadata = {worker_id: {}}
+
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:477]: I know this doesn't come from this PR, but still related: for every batch we also get the corresponding state, right? Have you seen how large they are? Asking because they will also be transferred through shared memory. Does it slow down the process a lot? Probably not if you have some io-heavy stuff in your dataset, but what if returning preloaded data?

File: pytorch_lightning/utilities/auto_restart.py [hunk 13]
```diff
@@ -589,3 +621,106 @@ def state_dict(self) -> Dict[str, Any]:
...
+def _get_iterator(self) -> "_BaseDataLoaderIter":
+    if not hasattr(self, "_lightning_fetcher"):
+        raise MisconfigurationException(
+            "A stateful iterator should be used only when a DataFetcher has been attached to the DataLoader."
+        )
+    if self.num_workers == 0:
+        return _SingleProcessDataLoaderIterStateful(self)
+    else:
+        if hasattr(self, "check_worker_number_rationality"):
+            self.check_worker_number_rationality()
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:709]: What does this do and where does it come from? Is this the warning we print if people don't use the correct number of workers? If yes, why is it here? Don't we already do that somewhere else?

**Review by justusschock — APPROVED** [review]
LGTM!

One general and one minor question.

File: pytorch_lightning/utilities/auto_restart.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:477]: Yes, with the BoringModel, we get a 30 % slowdown. But with Resnet50 on Cifar10, it is close to 0. @awaelchli performed the benchmark.

File: pytorch_lightning/utilities/auto_restart.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:709]: This is coming from the DataLoader `get_iterators`. In 1.10, `check_worker_number_rationality` was added, but doesn't exist on all versions. As I am patching their function, I had to still support this functionality.

**Review by SeanNaren — APPROVED** [review]
Looks good. imo we should offer a mixin class to add `state_dict` an `load_state_dict` for samplers/datasets, primarily out of good practice. But this can come at a later date :)
